### ISSUE TITLE
Sampling from Parameterized Bernoulli Distribution

### ISSUE THREAD

**Issue by Barcavin** [Graph-COM/GSAT#9]
When generating the subgraph, GSAT samples from the `Bernoulli(sigmoid(att_log_logits))` to get the "soft" value of the Bernoulli samples (it seems to be a continuous value rather than discrete 1/0). 

I don't quite understand how Gumbel-Softmax is applied here (https://github.com/Graph-COM/GSAT/blob/b3d798fe53f13509a74f02263b5d48e09fde1cd4/src/run_gsat.py#L308). I don't think this function make it sample from Gumbel distribution. It should be `-log(-log(u))` and u is from uniform(0,1) to be from Gumbel's. How does Gumbel softmax is applied?

**Comment by siqim** [comment]
Hi,

Thanks for your interest in our work. We indeed utilize the Gumbel-softmax trick to generate "soft" values as otherwise this part of the model would be non-differentiable. Regarding the implementation of the Gumbel-softmax trick with Bernoulli distributions, check out Appendix B and C.3.2 in https://arxiv.org/pdf/1611.00712.pdf.

Best,
Siqi

**Comment by Barcavin** [comment]
Thanks for the clarification!

In the ReadMe, you also mentioned that it is not appropriate to generate the hard sample thanks to MPNNs, so the soft values are used as normalizers for the message passing. However, I think the hard sample may also work by sampling a hard version of $\alpha$, where each entry is 0 or 1 instead of continuous values.

**Comment by siqim** [comment]
Yes, you are correct. With Gumbel-softmax, one can further use the straight-through trick mentioned in https://arxiv.org/pdf/1611.01144.pdf to generate "differentiable" discrete Bernoulli masks for message passing. The statement mentioned in the readme meant to state the operation of $\alpha \odot A$ is non-trivial to be applied appropriately for general MPNNs even with $\alpha$ being sampled using Gumbel-softmax, e.g., many MPNNs don't really use the adjacent matrix `A` and matrix multiplications to pass/aggr messages instead they use sparse `edge_index` and scatter operations, in which $\alpha \odot A$ can be hard to apply.

**Comment by Barcavin** [comment]
That makes sense to me.

It actually raises another question. How about sampling discrete $\alpha$ and multiplying them by `edge_index` in the `message()` function? In that case, general GNNs can then be combined with hard sample values.

**Comment by siqim** [comment]
This may be hard to implement in a differentiable way using current PyG. As PyG uses scatter operations to aggregate messages, see https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html, in which `edge_index` is non-differentiable. If one use $\alpha$ to get a new `edge_index` in `message()`, the gradient may not flow well I think.

**Comment by Barcavin** [comment]
But in the current implementation, the sampled attention is also applied to the messages (which is `x`) rather than `edge_index`. So for the "hard" version, the gradient flow can still get through in the same way as the "soft" version.

I am thinking a valid implementation might be to change https://github.com/Graph-COM/GSAT/blob/b3d798fe53f13509a74f02263b5d48e09fde1cd4/src/run_gsat.py#L307-L315 to

```python
    @staticmethod
    def concrete_sample(att_log_logit, temp, training):
        if training:
            random_noise = torch.empty_like(att_log_logit).uniform_(1e-10, 1 - 1e-10)
            random_noise = torch.log(random_noise) - torch.log(1.0 - random_noise)
            att_bern = (att_log_logit + random_noise) / temp
        else:
            att_bern = att_log_logit
        y_hard = (att_bern > 0).float()
        att_bern = y_hard - att_bern.detach() + att_bern
        return att_bern
```

**Comment by siqim** [comment]
Sorry, I thought you were saying using $\alpha$ to get a new masked `edge_index` instead of using $\alpha$ to mask the messages.  

If you were referring to mask messages with hard masks, then yes, by setting a proper threshold, with the straight-through trick one can mask the messages with "hard" masks, but if I remember it correctly, I tried it and that doesn't really bring a lot of benefits.

**Comment by Barcavin** [comment]
Thank you so much for your replying.  The paper itself is super interesting and the code is also neat and elegant. Thanks for publishing it!

**Comment by siqim** [comment]
Thank you for your interest in our work! Just let us know if you have any questions :)
